src/models/alpha_zero_model.py

```diff
@@ -0,0 +1,234 @@
+        self.policy = nn.Sequential(
+            nn.Linear(conv_policy_size, policy_shape)
+        ).to(device)
+
+        #optimizer
+        self.optimizer = optim.SGD(self.parameters(), lr=0.2, momentum=0.9, weight_decay=1e-4)
+
+
+    def _get_conv_val_size(self, shape):
+        o = self.conv_val(torch.zeros(1, *shape).to(self.device))
+        return int(np.prod(o.size()))
+
+    def _get_conv_policy_size(self, shape):
+        o = self.conv_policy(torch.zeros(1, *shape).to(self.device))
+        return int(np.prod(o.size()))
+
+    def forward(self, x):
+        batch_size = x.size()[0]
+
+        v = self.conv_in(x)
+        for net in self.residual_nets:
+            v = net(v)
+
+        val = self.conv_val(v)
+        val = self.value(val.view(batch_size, -1))
+
+        pol = self.conv_policy(v)
+        pol = self.policy(pol.view(batch_size, -1))
+
+        return pol, val
+
+    def predict(self, s, valids_actions):
+        """
+        Infer node data given an state
+
+        param s: 
+            list of encoded states of the game
+        param available_actions
+            list of encoded valids actions from a position of the game
+
+        return
+            (Move probabilities P vector, value V vector)
+        """
+        self.eval()
+        batch = self.state_lists_to_batch(s)
+        masks = [self.valids_actions_to_tensor(va) for va in valids_actions]
+        pol, val = self(batch)
+        pol = [self.get_policy_value(p, mask, False) for p, mask in zip(pol, masks)]
+        return pol, val
+
+    def get_policy_value(self, logits, valids_actions, log_softmax):
+        """
+        Get move probabilities distribution.
+
+        param logits:
+            list of 111 bits. Raw policy head
+        param available_actions
+            list of 111 bits. Mask of available actions
+        param log_softmax
+            Set True to use log_softmax as activation function. Set False to use softmax 
+
+        return
+            Move probabilities
+        """
+        mask = torch.tensor(valids_actions, dtype=torch.bool).to(self.device)
+        selection = torch.masked_select(logits, mask)
+        dist = F.log_softmax(selection, dim=-1)
+        if log_softmax:
+            return dist
+        return torch.exp(dist)
+
+    def state_lists_to_batch(self, state_lists):
+        """
+        Convert list of list states to batch for network
+
+        param state_lists: 
+            list of 'list[endoded states]'
+
+        return 
+            States to Tensor
+        """
+        assert isinstance(state_lists, list)
+        batch_size = len(state_lists)
+        batch = np.zeros((batch_size,) + STATE_SHAPE, dtype=np.float32)
+        for idx, state in enumerate(state_lists):
+            batch[idx] = state_to_list(state, 2542)
+        return torch.tensor(batch).to(self.device)
+
+    def valids_actions_to_tensor(self, valids_actions):
+        mask = state_to_list(valids_actions, 111)
```

Comment:
I think that we need to parameterize this number (111) in some way

Reply:
That's a hard one. The same applies to 2542 above. These are numbers tied to STATE_SHAPE hardcoded value and policy_shape arg of init method. But are also tied to encoder implementation. I can set 111 and 2542 to be tied to STATE_SHAPE and policy_shape, but if these values change, then I think the encoder can't do anything about it and still will be encoding state as original STATE_SHAPE.

Reply:
let's skip this for now